### PR TITLE
Proxy Multipart Upload for Azure Blob Storage

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/10531?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10531/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10531
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Implement proxy multipart upload for Azure Blob Storage

Example:

```shell
mlflow ui --artifacts-destination wasbs://testcontainer@teststorageaccount.blob.core.windows.net
```

```python
import os
import mlflow
from mlflow.store.artifact.http_artifact_repo import HttpArtifactRepository

os.environ["MLFLOW_MULTIPART_UPLOAD_MINIMUM_FILE_SIZE"] = str(5 * 1024**2)
os.environ["MLFLOW_MULTIPART_UPLOAD_CHUNK_SIZE"] = str(5 * 1024**2)

mlflow.set_tracking_uri("http://127.0.0.1:5000")
artifact_uri = mlflow.get_artifact_uri()
repo = mlflow.artifacts.get_artifact_repository(artifact_uri)
assert isinstance(repo, HttpArtifactRepository)
repo._try_multipart_upload("test.file", "my-folder")
```

image

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [X] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Implement proxy multipart upload for Azure Blob Storage

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/artifact/azure_blob_artifact_repo.py [hunk 1]
```diff
@@ -178,3 +197,62 @@ def _download_file(self, remote_file_path, local_path):
...
+        for part in parts:
+            qs = urllib.parse.urlparse(part.url).query
+            block_id = urllib.parse.parse_qs(qs)["blockid"][0]
+            block_id = decode_base64(urllib.parse.unquote(block_id))
+            block_ids.append(block_id)
+        blob_client = self.client.get_blob_client(container, dest_path)
+        blob_client.commit_block_list(block_ids)
+
+    def abort_multipart_upload(self, local_file, upload_id, artifact_path=None):
+        # There is no way to delete uncommitted blocks in Azure Blob Storage.
```
Comment (BenWilson2) [inline mlflow/store/artifact/azure_blob_artifact_repo.py:254]: great comment

File: mlflow/store/artifact/azure_blob_artifact_repo.py [hunk 2]
```diff
@@ -178,3 +197,62 @@ def _download_file(self, remote_file_path, local_path):
...
+        # Put Block: https://learn.microsoft.com/en-us/rest/api/storageservices/put-block?tabs=microsoft-entra-id
+        # SDK: https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobclient?view=azure-python#azure-storage-blob-blobclient-stage-block
+        blob_url = posixpath.join(self.client.url, container, dest_path)
+        sas_token = generate_blob_sas(
+            account_name=self.client.account_name,
+            container_name=container,
+            blob_name=dest_path,
+            account_key=self.client.credential.account_key,
+            permission=BlobSasPermissions(read=True, write=True),
+            expiry=datetime.datetime.utcnow() + datetime.timedelta(hours=6),
```
Comment (BenWilson2) [inline mlflow/store/artifact/azure_blob_artifact_repo.py:218]: Curious about the expiration time - is this something that Azure recommends for duration?

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for knocking this out @gabrielfu :D

File: mlflow/store/artifact/azure_blob_artifact_repo.py (hunk 2 above)
Comment (gabrielfu) [reply mlflow/store/artifact/azure_blob_artifact_repo.py:218]: I think I misread the documentation somewhere. Let me change to 1 hour to align with our s3 & gcs impl

**Comment by BenWilson2** [comment]
@gabrielfu Let's wait until after the 2.9 release to merge this :D

**Comment by gabrielfu** [comment]
> @gabrielfu Let's wait until after the 2.9 release to merge this :D

Sure!

**Pull request merged**